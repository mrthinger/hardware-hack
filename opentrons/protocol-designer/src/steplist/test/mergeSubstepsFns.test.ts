import { it, describe, expect } from 'vitest'
import {
  mergeSubstepRowsSingleChannel,
  mergeSubstepRowsMultiChannel,
} from '../generateSubstepItem'

const ingred1Id = 'ingred1Id'
const wellNamesForCol = (isMulti: boolean, colNum: string): string[] =>
  isMulti ? 'ABCDEFGH'.split('').map(s => `${s}${colNum}`) : [`A${colNum}`]

interface Ingreds {
  [ingredId: string]: number
}
const repeatIngreds = (
  isMulti: boolean,
  colNum: string,
  _ingreds: Ingreds | null | undefined
): Ingreds | Ingreds[] => {
  const ingreds = _ingreds || {}
  return isMulti
    ? wellNamesForCol(true, colNum).reduce(
        (acc, wellName) => ({ ...acc, [wellName]: ingreds }),
        {}
      )
    : ingreds
}

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const getFixtures = ({ isMulti }: { isMulti: boolean }) => {
  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
  const makeIngreds = (volume: number | null, colNum: string) =>
    repeatIngreds(isMulti, colNum, volume ? { [ingred1Id]: volume } : null)
  // NOTE: these cases do not cover dynamic behavior of `activeTips` key
  const activeTips = { labware: 'someTiprackId', well: 'A6' }
  return {
    activeTips,
    transferRowsFixture: [
      // from 'aspirate' command
      {
        activeTips,
        source: {
          wells: wellNamesForCol(isMulti, '1'),
          preIngreds: makeIngreds(30, '1'),
          postIngreds: makeIngreds(20, '1'),
        },
        volume: 10,
      },
      // from 'dispense' command
      {
        activeTips,
        dest: {
          wells: wellNamesForCol(isMulti, '12'),
          preIngreds: makeIngreds(null, '12'),
          postIngreds: makeIngreds(10, '12'),
        },
        volume: 10,
      },
    ],
    consolidateRowsFixture: [
      // from 'aspirate' command
      {
        activeTips,
        source: {
          wells: ['A1'],
          preIngreds: makeIngreds(30, '1'),
          postIngreds: makeIngreds(25, '1'),
        },
        volume: 5,
      },
      // second 'aspirate'
      {
        activeTips,
        source: {
          wells: wellNamesForCol(isMulti, '2'),
          preIngreds: makeIngreds(36, '2'),
          postIngreds: makeIngreds(31, '2'),
        },
        volume: 5,
      },
      // from 'dispense' command
      {
        activeTips,
        dest: {
          wells: wellNamesForCol(isMulti, '12'),
          preIngreds: makeIngreds(null, '12'),
          postIngreds: makeIngreds(10, '12'),
        },
        volume: 10,
      },
    ],
    distributeRowsFixture: [
      // from 'aspirate' command
      {
        activeTips,
        source: {
          wells: wellNamesForCol(isMulti, '1'),
          preIngreds: makeIngreds(30, '1'),
          postIngreds: makeIngreds(20, '1'),
        },
        volume: 10,
      },
      // first 'dispense'
      {
        activeTips,
        dest: {
          wells: wellNamesForCol(isMulti, '11'),
          preIngreds: makeIngreds(null, '11'),
          postIngreds: makeIngreds(5, '11'),
        },
        volume: 5,
      },
      // second 'dispense'
      {
        activeTips,
        dest: {
          wells: wellNamesForCol(isMulti, '12'),
          preIngreds: makeIngreds(null, '12'),
          postIngreds: makeIngreds(5, '12'),
        },
        volume: 5,
      },
    ],
  }
}

describe('mergeSubstepRowsSingleChannel', () => {
  const {
    activeTips,
    transferRowsFixture,
    consolidateRowsFixture,
    distributeRowsFixture,
  } = getFixtures({ isMulti: false })

  const testCases: Array<{
    testName: string
    showDispenseVol: boolean
    substepRows: any
    expected: any
  }> = [
    {
      testName: 'mock transfer / mix',
      showDispenseVol: false,
      substepRows: transferRowsFixture,
      expected: [
        // merged into single row
        {
          activeTips,
          source: {
            well: 'A1',
            preIngreds: { [ingred1Id]: 30 },
            postIngreds: { [ingred1Id]: 20 },
          },
          dest: {
            well: 'A12',
            preIngreds: {},
            postIngreds: { [ingred1Id]: 10 },
          },
          volume: 10,
        },
      ],
    },
    {
      testName: 'mock consolidate',
      showDispenseVol: false,
      substepRows: consolidateRowsFixture,
      expected: [
        // first aspirate stands alone
        {
          activeTips,
          source: {
            well: 'A1',
            preIngreds: { [ingred1Id]: 30 },
            postIngreds: { [ingred1Id]: 25 },
          },
          volume: 5,
        },
        // last asp + disp merged into single row
        {
          activeTips,
          source: {
            well: 'A2',
            preIngreds: { [ingred1Id]: 36 },
            postIngreds: { [ingred1Id]: 31 },
          },
          dest: {
            well: 'A12',
            preIngreds: {},
            postIngreds: { [ingred1Id]: 10 },
          },
          volume: 5,
        },
      ],
    },
    {
      testName: 'mock distribute',
      showDispenseVol: true, // IRL, this is only true for distribute
      substepRows: distributeRowsFixture,
      expected: [
        // first aspirate + disp merged into single row
        {
          activeTips,
          source: {
            well: 'A1',
            preIngreds: { [ingred1Id]: 30 },
            postIngreds: { [ingred1Id]: 20 },
          },
          dest: {
            well: 'A11',
            preIngreds: {},
            postIngreds: { [ingred1Id]: 5 },
          },
          volume: 5,
        },
        // last asp stands alone
        {
          activeTips,
          dest: {
            well: 'A12',
            preIngreds: {},
            postIngreds: { [ingred1Id]: 5 },
          },
          volume: 5,
        },
      ],
    },
  ]
  testCases.forEach(({ testName, showDispenseVol, substepRows, expected }) => {
    it(testName, () => {
      const result = mergeSubstepRowsSingleChannel({
        substepRows,
        showDispenseVol,
      })
      expect(result).toEqual(expected)
    })
  })
})

describe('mergeSubstepRowsMultiChannel', () => {
  const {
    transferRowsFixture,
    consolidateRowsFixture,
    distributeRowsFixture,
  } = getFixtures({ isMulti: true })
  const testCases: Array<{
    testName: string
    showDispenseVol: boolean
    isMixStep: boolean
    substepRows: any
  }> = [
    {
      testName: 'mock transfer',
      showDispenseVol: false,
      isMixStep: false,
      substepRows: transferRowsFixture,
    },
    {
      testName: 'mock mix',
      showDispenseVol: false,
      isMixStep: true,
      substepRows: transferRowsFixture,
    },
    {
      testName: 'mock consolidate',
      showDispenseVol: false,
      isMixStep: false,
      substepRows: consolidateRowsFixture,
    },
    {
      testName: 'mock distribute',
      showDispenseVol: true, // NOTE: IRL, should only be true for distribute
      isMixStep: false,
      substepRows: distributeRowsFixture,
    },
  ]
  testCases.forEach(({ testName, isMixStep, showDispenseVol, substepRows }) => {
    it(testName, () => {
      const channels = 8
      const result = mergeSubstepRowsMultiChannel({
        channels,
        showDispenseVol,
        isMixStep,
        substepRows,
      })
      expect(result).toMatchSnapshot()
    })
  })
})
