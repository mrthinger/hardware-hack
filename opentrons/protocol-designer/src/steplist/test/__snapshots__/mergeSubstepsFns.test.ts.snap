// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mergeSubstepRowsMultiChannel > mock consolidate 1`] = `
[
  [
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": undefined,
      "source": {
        "postIngreds": {
          "ingred1Id": 25,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "A1",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": undefined,
      "source": {
        "postIngreds": undefined,
        "preIngreds": undefined,
        "well": undefined,
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": undefined,
      "source": {
        "postIngreds": undefined,
        "preIngreds": undefined,
        "well": undefined,
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": undefined,
      "source": {
        "postIngreds": undefined,
        "preIngreds": undefined,
        "well": undefined,
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": undefined,
      "source": {
        "postIngreds": undefined,
        "preIngreds": undefined,
        "well": undefined,
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": undefined,
      "source": {
        "postIngreds": undefined,
        "preIngreds": undefined,
        "well": undefined,
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": undefined,
      "source": {
        "postIngreds": undefined,
        "preIngreds": undefined,
        "well": undefined,
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": undefined,
      "source": {
        "postIngreds": undefined,
        "preIngreds": undefined,
        "well": undefined,
      },
      "volume": 5,
    },
  ],
  [
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "A12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 31,
        },
        "preIngreds": {
          "ingred1Id": 36,
        },
        "well": "A2",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "B12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 31,
        },
        "preIngreds": {
          "ingred1Id": 36,
        },
        "well": "B2",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "C12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 31,
        },
        "preIngreds": {
          "ingred1Id": 36,
        },
        "well": "C2",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "D12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 31,
        },
        "preIngreds": {
          "ingred1Id": 36,
        },
        "well": "D2",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "E12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 31,
        },
        "preIngreds": {
          "ingred1Id": 36,
        },
        "well": "E2",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "F12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 31,
        },
        "preIngreds": {
          "ingred1Id": 36,
        },
        "well": "F2",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "G12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 31,
        },
        "preIngreds": {
          "ingred1Id": 36,
        },
        "well": "G2",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "H12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 31,
        },
        "preIngreds": {
          "ingred1Id": 36,
        },
        "well": "H2",
      },
      "volume": 5,
    },
  ],
]
`;

exports[`mergeSubstepRowsMultiChannel > mock distribute 1`] = `
[
  [
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "A11",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "A1",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "B11",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "B1",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "C11",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "C1",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "D11",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "D1",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "E11",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "E1",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "F11",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "F1",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "G11",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "G1",
      },
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "H11",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "H1",
      },
      "volume": 5,
    },
  ],
  [
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "A12",
      },
      "source": undefined,
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "B12",
      },
      "source": undefined,
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "C12",
      },
      "source": undefined,
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "D12",
      },
      "source": undefined,
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "E12",
      },
      "source": undefined,
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "F12",
      },
      "source": undefined,
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "G12",
      },
      "source": undefined,
      "volume": 5,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 5,
        },
        "preIngreds": {},
        "well": "H12",
      },
      "source": undefined,
      "volume": 5,
    },
  ],
]
`;

exports[`mergeSubstepRowsMultiChannel > mock mix 1`] = `
[
  [
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "A1",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "A1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "B1",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "B1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "C1",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "C1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "D1",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "D1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "E1",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "E1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "F1",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "F1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "G1",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "G1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "H1",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "H1",
      },
      "volume": 10,
    },
  ],
]
`;

exports[`mergeSubstepRowsMultiChannel > mock transfer 1`] = `
[
  [
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "A12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "A1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "B12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "B1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "C12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "C1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "D12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "D1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "E12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "E1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "F12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "F1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "G12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "G1",
      },
      "volume": 10,
    },
    {
      "activeTips": {
        "labware": "someTiprackId",
        "well": "A6",
      },
      "dest": {
        "postIngreds": {
          "ingred1Id": 10,
        },
        "preIngreds": {},
        "well": "H12",
      },
      "source": {
        "postIngreds": {
          "ingred1Id": 20,
        },
        "preIngreds": {
          "ingred1Id": 30,
        },
        "well": "H1",
      },
      "volume": 10,
    },
  ],
]
`;
